crates/modelardb_server/src/storage/types.rs

```diff
@@ -0,0 +1,291 @@
+    /// is saved to disk and if the amount of memory for uncompressed data is reduced.
+    remaining_compressed_memory_in_bytes: RwLock<isize>,
+}
+
+/// The pool of memory the
+/// [`UncompressedDataManager`](crate::storage::uncompressed_data_manager::UncompressedDataManager)
+/// can use for uncompressed data and the
+/// [`CompressedDataManager`](crate::storage::CompressedDataManager) can use for compressed data.
+impl MemoryPool {
+    /// Create a new [`MemoryPool`] with at most [`u64::MAX`] bytes of memory for uncompressed data
+    /// and at most [`i64::MAX`] bytes of memory for compressed data.
+    pub(super) fn new(
+        uncompressed_memory_in_bytes: usize,
+        compressed_memory_in_bytes: usize,
+    ) -> Self {
+        // unwrap() is safe as i64::MAX is 8192 PiB and the value is from ConfigurationManager.
+        Self {
+            remaining_uncompressed_memory_in_bytes: RwLock::new(
+                uncompressed_memory_in_bytes.try_into().unwrap(),
+            ),
+            remaining_compressed_memory_in_bytes: RwLock::new(
+                compressed_memory_in_bytes.try_into().unwrap(),
+            ),
+        }
+    }
+
+    /// Change the amount of memory available for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata by `size_in_bytes`.
+    pub(super) fn update_uncompressed_memory(&self, size_in_bytes: isize) {
+        // unwrap() is safe as write() only returns an error if the lock is poisoned.
+        *self.remaining_uncompressed_memory_in_bytes.write().unwrap() += size_in_bytes
+    }
+
+    /// Return the amount of memory available for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata in bytes.
+    pub(super) fn remaining_uncompressed_memory_in_bytes(&self) -> isize {
+        // unwrap() is safe as write() only returns an error if the lock is poisoned.
+        *self.remaining_uncompressed_memory_in_bytes.read().unwrap()
+    }
+
+    /// Try to reserve `size_in_bytes` bytes of memory for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata. Returns [`true`] if the reservation
+    /// succeeds and [`false`] otherwise.
+    pub(super) fn try_reserve_uncompressed_memory(&self, size_in_bytes: usize) -> bool {
```

Comment:
Does it break Rust conventions to use `try_` with a function that return `bool` and not a `Result`? Personally I do not mind. 

If it is changed here, also change it in the other method.

Reply:
While `try_` generally indicate that a method can fail and most return a `Result`, I am not aware of any convention that it should return `Result`. I used `bool` instead of `Result` as it made the API nicer to use in my opinion, but we can change to an error type that indicates that memory could not be reserved and then use `if ....is_err()` if you think it is beneficial?

Reply:
As it does not go against any conventions I also think the current API is fine.